### PR TITLE
Udp shim INT 2.0+

### LINKED ISSUE #87: Add support for UDP shim per INT spec with probe

### LINKED ISSUE #118: Add IPv6 with INT support to LUA plugin
Ensure IPv6 packets with INT can be parsed and displayed properly in Wireshark.

### LINKED ISSUE #121: AE crashes when receiving IPv6 INT packets
With the IPv6 support that PR #120 adds, any INT IPv6 packets getting received by the AE will cause it to crash; however, the unit tests in oinc_tests.py mocking up the expected INT packets does not cause any exceptions.

### LINKED ISSUE #127: Add support for UDP encapsulation in wireshark plugin
To be completed either during or after working on issue #87

### PR DESCRIPTION
**Pull request by spisarski**
#### What does this PR do?
Updates the INT headers to reflect the P4 INT 2.0 specification plus the orig_mac value from the gateway.
Fixes #87 
Fixes #127 
Fixes #121 
Fixes #118 
#### Do you have any concerns with this PR?
I had to return the recirculation logic back to core.p4
#### How can the reviewer verify this PR?
Check CI status or run automation scripts
#### Questions:
- Have you connected this PR to the issue it resolves? yes
- Does the documentation need an update? no
- Does this add new dependencies? no
- Have you added unit or functional tests for this PR? Ensured they were written properly

### CODE REVIEW

**Comment by spisarski** [comment]
Please do not merge until Jenkins is back up and running the tests

**Comment by spisarski** [comment]
See the last attachment on issue #125 which is a zip file of packet dumps from each level.

**Comment by spisarski** [comment]
Run Tests

**Comment by spisarski** [comment]
Run Tests

**Comment by spisarski** [comment]
Run Tests

File: bin/full-gen-pcap-frg.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+sudo python /home/ubuntu/transparent-security/trans_sec/device_software/send_packets.py -y 0 -i 0.005 -it 1 -itd 0 -z gateway1-eth1 -sa 192.168.1.2 -r 10.2.5.2 -sp 9648 -p 3074 -c 1 -e 00:00:00:00:05:05 -s 00:00:00:01:01:00 -pr TCP -m 'TCP Normal IPv4'
+
+sudo python /home/ubuntu/transparent-security/trans_sec/device_software/send_packets.py -y 0 -i 0.005 -it 1 -itd 0 -z gateway1-eth1 -sa 0000:0000:0000:0000:0000:0001:0001:0002 -r 0000:0000:0000:0000:0000:0002:0001:0002 -sp 9648 -p 3074 -c 1 -e 00:00:00:00:05:05 -s 00:00:00:01:01:00 -pr UDP -m 'UDP Normal IPv6'
+sudo python /home/ubuntu/transparent-security/trans_sec/device_software/send_packets.py -y 0 -i 0.005 -it 1 -itd 0 -z gateway1-eth1 -sa 0000:0000:0000:0000:0000:0001:0001:0002 -r 0000:0000:0000:0000:0000:0002:0001:0002 -sp 9648 -p 3074 -c 1 -e 00:00:00:00:05:05 -s 00:00:00:01:01:00 -pr TCP -m 'TCP Normal IPv6'
+
+
+# Commands for sniffing the important interfaces while generating the packets above
```
Comment (RandyLevensalor) [inline bin/full-gen-pcap-frg.sh]: Why not start tcpdump before generating the packets?

Also filter the packets.  Something simple like udp or tcp would be a start, to just capture the packets that you generate.
sudo tcpdump -i gateway1-eth1 udp or tcp -w gateway1-eth1.pcap

File: docs/int_header/int.lua [hunk 2]
```diff
@@ -1,98 +1,151 @@
...
-tps_proto = Proto("TPS_INT","TPS INT Protocol")
--- create a function to dissect it
-function tps_proto.dissector(buffer,pinfo,tree)
-    pinfo.cols.protocol = "TPS INT"
+function tps_int_shim(int_tree, shim_buf)
+    -- UDP INT Shim Header - 4 bytes
+    local shim_tree = int_tree:add(shim_buf, "UDP INT Shim Header")
+    shim_tree:add("Type: " .. shim_buf:bitfield(0, 4))
+    shim_tree:add("NPT: " .. shim_buf:bitfield(4, 2))
+    assert(shim_buf:bitfield(6, 2)) -- reserved
```
Comment (RandyLevensalor) [inline docs/int_header/int.lua]: I thought that the assertion check would be asserting that the values are == nil or 0?

Add an error message with each assert.

I think that it should be something like:
local reserved_1 = shim_buf:bitfield(6, 2)
assert(reserved_1 == 0, "Reserved bits 6-8 in UDP INT Shim header are '" + reserved_1 + "' expected '0'")

File: p4/include/tps_checksum.p4 [hunk 3]
```diff
@@ -19,29 +19,69 @@
...
+	        hdr.ipv4.ihl,
+                hdr.ipv4.diffserv,
+                hdr.ipv4.totalLen,
+                hdr.ipv4.identification,
+                hdr.ipv4.flags,
+                hdr.ipv4.fragOffset,
+                hdr.ipv4.ttl,
+                hdr.ipv4.protocol,
+                hdr.ipv4.srcAddr,
+                hdr.ipv4.dstAddr
```
Comment (RandyLevensalor) [inline p4/include/tps_checksum.p4]: Spacing is off.  maybe tabs and not spaces?

File: p4/include/tps_checksum.p4 [hunk 4]
```diff
@@ -19,29 +19,69 @@
...
```
Comment (RandyLevensalor) [inline p4/include/tps_checksum.p4]: Indentation is off.

Could this be done with a struct or something to eliminate all of the duplicate lines?

File: p4/include/tps_consts.p4 [hunk 5]
```diff
@@ -13,12 +13,30 @@
...
+const bit<2> INT_SHIM_NPT_UDP_FULL_WRAP = 0x2;
+const bit<4> INT_VERSION = 0x2;
+const bit<5> INT_META_LEN = 0x1;
+const bit<16> UDP_INT_SRC_PORT = 0x0;
+const bit<16> UDP_INT_DST_PORT = 0x022b;
+const bit<16> TPS_UDP_PORT = 0x216b;
 const bit<16> TYPE_IPV4 = 0x0800;
 const bit<16> TYPE_IPV6 = 0x86dd;
 const bit<8> TYPE_TCP = 0x06;
 const bit<8> TYPE_UDP = 0x11;
```
Comment (RandyLevensalor) [inline p4/include/tps_consts.p4]: Add comments for defines and constants. What is it used for and why was that value selected.

Perhaps more descriptive names for some of the generic names.

File: p4/include/tps_consts.p4 [hunk 6]
```diff
@@ -13,12 +13,30 @@
...
+
```
Comment (RandyLevensalor) [inline p4/include/tps_consts.p4]: Why a blank line here?  If it is a separate logical block, please put a comment before each block.

File: p4/include/tps_headers.p4 [hunk 7]
```diff
@@ -134,14 +140,15 @@ struct metadata {
...
+    int_header_t      int_header;
+    int_metadata_t    int_meta_3;
+    int_metadata_t    int_meta_2;
     source_metadata_t int_meta;
-    int_metadata_t int_meta_2;
-    int_metadata_t int_meta_3;
-    udp_t          udp;
-    tcp_t          tcp;
+    udp_t             udp;
+    tcp_t             tcp;
```
Comment (RandyLevensalor) [inline p4/include/tps_headers.p4]: Why parse the tcp header?

We should just need to know the IP protocol for all non UDP packets.

File: p4/gateway/gateway.p4 [hunk 8]
```diff
@@ -106,10 +124,34 @@ control TpsGwIngress(inout headers hdr,
...
+        }
+        size = TABLE_SIZE;
+        default_action = insert_udp_int_for_udp();
+    }
+
+    action insert_udp_int_for_tcp() {
+        hdr.udp_int.len = ((bit<16>)hdr.int_shim.length * BYTES_PER_SHIM * INT_SHIM_HOP_SIZE) + TCP_HDR_BYTES + UDP_HDR_BYTES;
+    }
+
+    table insert_udp_int_for_tcp_t {
```
Comment (RandyLevensalor) [inline p4/gateway/gateway.p4]: Why is this specific for TCP and not just general for all none UDP traffic?

File: p4/gateway/gateway.p4 [hunk 9]
```diff
@@ -53,46 +53,64 @@ control TpsGwIngress(inout headers hdr,
...
-        hdr.ipv4.totalLen = hdr.ipv4.totalLen + ((bit<16>)hdr.int_shim.length * 4);
+        hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
+        hdr.ipv4.protocol = TYPE_UDP;
+        hdr.ipv4.totalLen = hdr.ipv4.totalLen + ((bit<16>)hdr.int_shim.length * BYTES_PER_SHIM * INT_SHIM_HOP_SIZE) + UDP_HDR_BYTES;
+
         forwardedPackets.count(device);
     }
 
     action data_inspect_packet_ipv6(bit<32> device, bit<32> switch_id) {
+        hdr.udp_int.setValid();
```
Comment (RandyLevensalor) [inline p4/gateway/gateway.p4]: Could we check that this packet doesn't already have a UDP Shim header before we add one?

File: bin/full-gen-pcap-frg.sh (hunk 1 above)
Comment (spisarski) [reply bin/full-gen-pcap-frg.sh]: Was never meant to be executable

File: docs/int_header/int.lua (hunk 2 above)
Comment (spisarski) [reply docs/int_header/int.lua]: I tried that but lua didn't like it

File: p4/include/tps_checksum.p4 (hunk 4 above)
Comment (spisarski) [reply p4/include/tps_checksum.p4]: Didn't notice them earlier but I've already have that fix with the telem report work I'm doing.

File: p4/include/tps_headers.p4 (hunk 7 above)
Comment (spisarski) [reply p4/include/tps_headers.p4]: for gateway

File: p4/gateway/gateway.p4 (hunk 8 above)
Comment (spisarski) [reply p4/gateway/gateway.p4]: need to add the different header lengths to the udp_int.len value

File: p4/gateway/gateway.p4 (hunk 9 above)
Comment (spisarski) [reply p4/gateway/gateway.p4]: The current logic is that gateways are not looking for or parsing INT headers. If that is something we need to support, please create a ticket.

File: p4/include/tps_checksum.p4 (hunk 3 above)
Comment (spisarski) [reply p4/include/tps_checksum.p4]: Found and fixed with the telemetry report stuff. I'll fix this later.

File: p4/include/tps_consts.p4 (hunk 5 above)
Comment (spisarski) [reply p4/include/tps_consts.p4]: I'll push in the comments soon.

File: p4/include/tps_consts.p4 (hunk 6 above)
Comment (spisarski) [reply p4/include/tps_consts.p4]: Will remove erroneous lines with the comments

File: p4/gateway/gateway.p4 (hunk 8 above)
Comment (spisarski) [reply p4/gateway/gateway.p4]: I am however going to remove the tables and just call the action directly from apply

**Pull request merged**